Source/WebCore/bindings/js/JSNodeListCustom.cpp

```diff
@@ -37,6 +37,84 @@
 namespace WebCore {
 using namespace JSC;
 
+// FIXME: Instead of handpicking methods from JSNodeList / JSObject, extract JSNodeListBase and make
+// JSStaticNodeList / JSLiveNodeList extend it (which would require massive changes to CodeGeneratorJS.pm though).
+const ClassInfo JSStaticNodeList::s_info = {
+    "NodeList"_s,
+    &Base::s_info,
+    nullptr,
+    nullptr,
+    std::nullopt, {
+        &JSStaticNodeList::destroy,
+        &JSStaticNodeList::getCallData,
+        &JSStaticNodeList::getConstructData,
+        &JSStaticNodeList::put,
+        &JSStaticNodeList::putByIndex,
+        &JSStaticNodeList::deleteProperty,
+        &JSStaticNodeList::deletePropertyByIndex,
+        &Base::Base::getOwnPropertySlot,
+        &Base::Base::getOwnPropertySlotByIndex,
+        &JSStaticNodeList::toThis,
+        &Base::Base::getOwnPropertyNames,
+        &Base::Base::getOwnSpecialPropertyNames,
+        &JSStaticNodeList::customHasInstance,
+        &JSStaticNodeList::defineOwnProperty,
+        &JSStaticNodeList::preventExtensions,
+        &JSStaticNodeList::isExtensible,
+        &JSStaticNodeList::setPrototype,
+        &JSStaticNodeList::getPrototype,
+        &JSStaticNodeList::dumpToStream,
+        &JSStaticNodeList::analyzeHeap,
+        &JSStaticNodeList::estimatedSize,
+        &JSStaticNodeList::visitChildren,
+        &JSStaticNodeList::visitChildren,
+        &JSStaticNodeList::visitOutputConstraints,
+        &JSStaticNodeList::visitOutputConstraints,
+    },
+    sizeof(JSStaticNodeList),
+};
+
+JSObject* JSStaticNodeList::createPrototype(VM& vm, JSDOMGlobalObject& globalObject)
+{
+    return Base::prototype(vm, globalObject);
+}
+
+JSStaticNodeList::JSStaticNodeList(Structure* structure, JSDOMGlobalObject& globalObject, Ref<StaticNodeList>&& impl)
+    : Base(structure, globalObject, WTFMove(impl))
+{
+}
+
+void JSStaticNodeList::finishCreation(VM& vm)
+{
+    Base::finishCreation(vm);
+
+    auto& nodeList = wrapped();
+    unsigned length = nodeList.length();
+
+    IndexingHeader indexingHeader;
+    indexingHeader.setVectorLength(length);
+    indexingHeader.setPublicLength(length);
+    auto* butterfly = Butterfly::tryCreate(vm, nullptr, 0, structure()->outOfLineCapacity(), true, indexingHeader, sizeof(EncodedJSValue) * length);
+    if (UNLIKELY(!butterfly))
+        return;
+
+    auto* globalObject = this->globalObject();
+    auto contiguousValues = butterfly->contiguous();
+
+    for (unsigned i = 0; i < length; ++i)
+        contiguousValues.atUnsafe(i).clear();
+
+    setButterfly(vm, butterfly);
+
+    for (unsigned i = 0; i < length; ++i) {
+        auto jsNode = toJS(globalObject, globalObject, *nodeList.item(i));
+        contiguousValues.atUnsafe(i).setWithoutWriteBarrier(jsNode);
```

Comment:
We need to be emitting a barrier here since toJS can allocate.

Reply:
Replaced with `.set(`, thanks!